### ISSUE TITLE
Mutually recursive types issue

### ISSUE THREAD

**Issue by CkuT** [NoRedInk/elm-decode-pipeline#26]
Hi !

I am trying to have decode a JSON content which has mutually recursive types.

Here is a gist with the working version (**without** `decode-pipeline`) and with the non-working version (**with** `decode-pipeline`). I am very new to Elm, and I don't know if it is a `decode-pipeline` related issue, so I am sorry if it is not the case.

https://gist.github.com/CkuT/77d7bb080a9a70612fb4e74edbd34d12

With the non-working file, I get a `TypeError: decoder is undefined` at Runtime with `decode-pipeline`.

NB : Initially, I had the same problem (`decoder is undefined`) when I declared a function with a lazy decoder, but I didn't manage to reproduce it... It seems a bit dependent on the order declarations.
https://gist.github.com/CkuT/82d21c981bbae764f8f26a44983814ef

**Comment by jerith666** [comment]
I may be seeing something similar.  (Or I may be using https://github.com/agrafix/elm-bridge/ and/or https://github.com/bartavelle/json-helpers incorrectly?)

Here's the stack trace:

```
TypeError: decoder is undefined[Learn More]  dest:4588:1
	runHelp http://127.0.0.1/~mmchenry/sinb/test/dest/:4588:1
	runHelp http://127.0.0.1/~mmchenry/sinb/test/dest/:4721:18
	run http://127.0.0.1/~mmchenry/sinb/test/dest/:4580:15
	A2 http://127.0.0.1/~mmchenry/sinb/test/dest/:94:7
	_bartavelle$json_helpers$Json_Helpers$decodeSumFinal< http://127.0.0.1/~mmchenry/sinb/test/dest/:5969:11
	A4 http://127.0.0.1/~mmchenry/sinb/test/dest/:106:7
	_bartavelle$json_helpers$Json_Helpers$decodeSumObjectWithSingleField</< http://127.0.0.1/~mmchenry/sinb/test/dest/:5994:13
	_bartavelle$json_helpers$Json_Helpers$customDecoder</< http://127.0.0.1/~mmchenry/sinb/test/dest/:5977:15
	runHelp http://127.0.0.1/~mmchenry/sinb/test/dest/:4734:15
	run http://127.0.0.1/~mmchenry/sinb/test/dest/:4580:15
	runOnString http://127.0.0.1/~mmchenry/sinb/test/dest/:4575:9
	A2 http://127.0.0.1/~mmchenry/sinb/test/dest/:94:7
	_elm_lang$http$Http$expectJson/< http://127.0.0.1/~mmchenry/sinb/test/dest/:9655:11
	handleResponse http://127.0.0.1/~mmchenry/sinb/test/dest/:9515:15
	toTask/</< http://127.0.0.1/~mmchenry/sinb/test/dest/:9425:13
```

**Comment by jerith666** [comment]
Hm ... or maybe I'm hitting https://github.com/elm-lang/elm-compiler/issues/873?  Still investigating ...

**Comment by jerith666** [comment]
Yeah, putting in a couple of `Json.Decode.lazy` calls seems to have fixed it; see https://github.com/elm-lang/elm-compiler/blob/master/hints/bad-recursion.md for details.  Maybe the original issue is amenable to the same kind of work-around?

**Comment by mateusfccp** [comment]
@jerith666 Using `Json.Decode.lazy` worked for me. I think Elm should prevent it at compile time, as Elm compiler is much better at hinting the problem than inspecting what browser complains about js. I think this issue should become a feature request about this.

**Comment by jerith666** [comment]
It looks like they're tracking this issue in https://github.com/elm-lang/elm-compiler/issues/1377 now.

**Comment by rtfeldman** [comment]
This is a compiler issue that's been fixed in Elm 0.19, so I'm going to close this issue here. 😄
